### ISSUE TITLE
Rust should explain why a closure call requires a mutable borrow

### ISSUE THREAD

**Issue by Aaron1011** [rust-lang/rust#80313]
The following code:

```rust
fn main() {
    let mut my_var = false;
    let callback = || {
        my_var = true;
    };
    callback();
}
```

produces the following error:

```
error[E0596]: cannot borrow `callback` as mutable, as it is not declared as mutable
 --> src/main.rs:6:5
  |
3 |     let callback = || {
  |         -------- help: consider changing this to be mutable: `mut callback`
...
6 |     callback();
  |     ^^^^^^^^ cannot borrow as mutable

error: aborting due to previous error
```

The expression `callback()` tries to mutably borrow `callback` due to the mutation of `my_var`. However, the error message doesn't explain this.

We should add a note to the error message pointing to the span of the upvar that caused us to need the mutable borrow.

**Comment by camelid** [comment]
So it looks like this is the function that needs to be modified:

https://github.com/rust-lang/rust/blob/89886e69369b845463cf1b27abd536ce3adb7c78/compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs#L23-L30

It seems I might be able to get the information I need from `self.used_mut_upvars`?

**Comment by camelid** [comment]
Actually, `self.used_mut_upvars` won't work here because it's only used when we're borrow-checking the closure, not uses of it (`self.used_mut_upvars` is empty in this context).

**Comment by 1000teslas** [comment]
Is anyone going to work on this issue? If not, I want to work on this to help me learn how Rust works.

**Comment by Aaron1011** [comment]
@1000teslas Go ahead! If you need any help, feel free to reach out to me on Zulip or Discord

**Comment by camelid** [comment]
I was working on this, but I guess I should have pinged Aaron: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/finding.20mutated.20upvars.20in.20borrowck. Oh well :)

@1000teslas you can work on this, I have other things that I've been meaning to get to anyway.

**Comment by 1000teslas** [comment]
@Aaron1011 What is the intended error? Is it meant to point to my_var, similarly to room_ref in https://github.com/rust-lang/rust/issues/78938?

**Comment by Aaron1011** [comment]
Yes - I was thinking of displaying a message like "a mutable borrow is required because the closure modifies `my_var`", located at the usage of `my_var`. Additionally, we could add another note pointing to the definition of `my_var`.

**Comment by 1000teslas** [comment]
I'm not sure how given a `Location` or `Place` of a call, we can obtain the `Local` pointing to the definition for `self.body.local_decls`.

**Comment by Aaron1011** [comment]
@1000teslas I think the information you need already exists in `closure_kind_origins()`. 

You can see an example of retrieving this here: https://github.com/rust-lang/rust/blob/edeb631ad0cd6fdf31e2e31ec90e105d1768be28/compiler/rustc_mir/src/borrow_check/diagnostics/conflict_errors.rs#L295-L298

and of using it: https://github.com/rust-lang/rust/blob/edeb631ad0cd6fdf31e2e31ec90e105d1768be28/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L591-L612

I don't think you'll need to access the MIR `Local`

**Comment by JohnTitor** [comment]
Fixed by #81158
